src/cargo/core/compiler/compile_kind.rs

```diff
@@ -0,0 +1,90 @@
+use crate::core::compiler::BuildContext;
+use crate::core::{InternedString, Target};
+use crate::util::errors::{CargoResult, CargoResultExt};
+use serde::Serialize;
+use std::path::Path;
+
+/// Indicates whether an object is for the host architcture or the target architecture.
+///
+/// These will be the same unless cross-compiling.
+#[derive(PartialEq, Eq, Hash, Debug, Clone, Copy, PartialOrd, Ord, Serialize)]
+pub enum CompileKind {
+    Host,
+    Target(CompileTarget),
+}
+
+impl CompileKind {
+    pub fn is_host(&self) -> bool {
+        match self {
+            CompileKind::Host => true,
+            _ => false,
+        }
+    }
+
+    pub fn for_target(self, target: &Target) -> CompileKind {
+        // Once we start compiling for the `Host` kind we continue doing so, but
+        // if we are a `Target` kind and then we start compiling for a target
+        // that needs to be on the host we lift ourselves up to `Host`.
+        match self {
+            CompileKind::Host => CompileKind::Host,
+            CompileKind::Target(_) if target.for_host() => CompileKind::Host,
+            CompileKind::Target(n) => CompileKind::Target(n),
+        }
+    }
+
+    /// Returns a "short" name for this kind, suitable for keying off
+    /// configuration in Cargo or presenting to users.
+    pub fn short_name(&self, bcx: &BuildContext<'_, '_>) -> &str {
+        match self {
+            CompileKind::Host => bcx.host_triple().as_str(),
+            CompileKind::Target(target) => target.short_name(),
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Hash, Debug, Clone, Copy, PartialOrd, Ord, Serialize)]
+pub struct CompileTarget {
```

Comment:
This could use a brief doc comment explaining that it may be a target triple or path (absolute?) to a `.json` file.

Also, I'm curious, did you consider making it an enum to differentiate the two?  If it is an enum, it could pre-compute `rustc_target` so it is not repeatedly recomputed.

Reply:
Comment added! I didn't actually consider using an `enum`, but I was a bit overzealous in that I was trying to keep this the same size as an `InternedString`. There's definitely not a great reason for that though.

I tried to remove the `Copy` bound on this to store something like `PathBuf`, but unfortunately that would require a good deal of work I think.

I'm sort of fine either way, I think having it encapsulated is fine for now, but I don't mind either way. Would you prefer to switch to an `enum`?

Reply:
No, it was just a thought to simplify `short_name`, but I don't think it is necessary.